plugin/sampling/strategystore/adaptive/processor.go

```diff
@@ -0,0 +1,522 @@
+	return float64(count) / seconds
+}
+
+// calculateWeightedQPS calculates the weighted qps of the slice allQPS where weights are biased
+// towards more recent qps. This function assumes that the most recent qps is at the head of the slice.
+func (p *processor) calculateWeightedQPS(allQPS []float64) float64 {
+	if len(allQPS) == 0 {
+		return 0
+	}
+	weights := p.weightVectorCache.getWeights(len(allQPS))
+	var qps float64
+	for i := 0; i < len(allQPS); i++ {
+		qps += allQPS[i] * weights[i]
+	}
+	return qps
+}
+
+func (p *processor) prependServiceCache() {
+	p.serviceCache = append([]samplingCache{make(samplingCache)}, p.serviceCache...)
+	if len(p.serviceCache) > serviceCacheSize {
+		p.serviceCache = p.serviceCache[0:serviceCacheSize]
+	}
+}
+
+func (p *processor) calculateProbabilitiesAndQPS() (model.ServiceOperationProbabilities, model.ServiceOperationQPS) {
+	p.prependServiceCache()
+	retProbabilities := make(model.ServiceOperationProbabilities)
+	retQPS := make(model.ServiceOperationQPS)
+	svcOpQPS := p.throughputToQPS()
+	totalOperations := int64(0)
+	for svc, opQPS := range svcOpQPS {
+		if _, ok := retProbabilities[svc]; !ok {
+			retProbabilities[svc] = make(map[string]float64)
+		}
+		if _, ok := retQPS[svc]; !ok {
+			retQPS[svc] = make(map[string]float64)
+		}
+		for op, qps := range opQPS {
+			totalOperations++
+			avgQPS := p.calculateWeightedQPS(qps)
+			retQPS[svc][op] = avgQPS
+			retProbabilities[svc][op] = p.calculateProbability(svc, op, avgQPS)
+		}
+	}
+	p.operationsCalculatedGauge.Update(totalOperations)
+	return retProbabilities, retQPS
+}
+
+func (p *processor) calculateProbability(service, operation string, qps float64) float64 {
+	oldProbability := p.InitialSamplingProbability
+	// TODO: is this loop overly expensive?
+	p.RLock()
+	if opProbabilities, ok := p.probabilities[service]; ok {
+		if probability, ok := opProbabilities[operation]; ok {
+			oldProbability = probability
+		}
+	}
+	latestThroughput := p.throughputs[0].throughput
+	p.RUnlock()
+
+	usingAdaptiveSampling := p.isUsingAdaptiveSampling(oldProbability, service, operation, latestThroughput)
+	p.serviceCache[0].Set(service, operation, &samplingCacheEntry{
+		probability:   oldProbability,
+		usingAdaptive: usingAdaptiveSampling,
+	})
+
+	// Short circuit if the qps is close enough to targetQPS or if the service doesn't appear to be using
+	// adaptive sampling.
+	if p.withinTolerance(qps, p.TargetSamplesPerSecond) || !usingAdaptiveSampling {
+		return oldProbability
+	}
+	var newProbability float64
+	if floatEquals(qps, 0) {
+		// Edge case; we double the sampling probability if the QPS is 0 so that we force the service
+		// to at least sample one span probabilistically.
+		newProbability = oldProbability * 2.0
+	} else {
+		newProbability = p.probabilityCalculator.Calculate(p.TargetSamplesPerSecond, qps, oldProbability)
+	}
+	return math.Min(maxSamplingProbability, math.Max(p.MinSamplingProbability, newProbability))
```

Comment:
the `maxSamplingProbability` const is superfluous, this stmt would read better if you simply use `1.0`